_episodes/10-diagnostics.md

```diff
@@ -0,0 +1,504 @@
+>>         'exp': 'historical',
+>>         'filename': '~/recipe_python_20210305_131929/preproc/map/tas/CMIP5_CanESM2_Amon_historical_r1i1p1_tas_2000-2000.nc',
+>>         'frequency': 'mon',
+>>         'institute': ['CCCma'],
+>>         'long_name': 'Near-Surface Air Temperature',
+>>         'mip': 'Amon',
+>>         'modeling_realm': ['atmos'],
+>>         'preprocessor': 'select_january',
+>>         'project': 'CMIP5',
+>>         'recipe_dataset_index': 1,
+>>         'short_name': 'tas',
+>>         'standard_name': 'air_temperature',
+>>         'start_year': 2000,
+>>         'units': 'K',
+>>         'variable_group': 'tas'}]}
+>>```
+>>
+>> This is how we can access preprocessed data within our diagnostic.
+> {: .solution}
+{: .challenge}
+
+## Diagnostic computation
+
+After grouping and selecting data, we can read individual attributes (such as filename) 
+of each item. Here we have grouped the input data  by ``variables`` 
+so we loop over the variables (line 89-93). Following this, is a call to the 
+function ``compute_diagnostic`` (line 94). Let's have a look at the
+definition of this function in line 43 where the actual analysis on the data is done.
+
+Note that output from the ESMValCore preprocessor is in the form of NetCDF files.
+Here, ``compute_diagnostic`` uses
+[Iris](https://scitools-iris.readthedocs.io/en/latest/index.html) to read data
+from a netCDF file and performs an operation ``squeeze`` to remove any dimensions
+of length one. We can adapt this function to add our own analysis. As an
+example, here we calculate the bias using the average of the data using Iris cubes. 
+
+~~~python
+def compute_diagnostic(filename):
+    """Compute an example diagnostic."""
+    logger.debug("Loading %s", filename)
+    cube = iris.load_cube(filename)
+
+    logger.debug("Running example computation")
+    cube = iris.util.squeeze(cube)
+
+    # Calculate a bias using the average of data
+    cube.data = cube.core_data() - cube.data.mean()
+    return cube
+~~~
+
+> ## iris cubes
+>
+> Iris reads data from NetCDF files into data structures called
+> [cubes](https://scitools-iris.readthedocs.io/en/latest/userguide/iris_cubes.html).
+> The data in these cubes can be modified, combined with other cubes' data or
+> plotted.
+{: .callout}
+
+> ## Reading data using xarray
+>
+> Alternately, you can use [xarrays](http://xarray.pydata.org/en/stable/) to read the data
+> instead of Iris.
+>
+>> ## Answer
+>>
+>> First, import ``xarray`` package at the top of the script as:
+>>
+>>~~~python
+>>import xarray as xr
+>>~~~
+>>
+>> Then, change the ``compute_diagnostic`` as:
+>>
+>>~~~python
+>>def compute_diagnostic(filename):
+>>    """Compute an example diagnostic."""
+>>    logger.debug("Loading %s", filename)
+>>    dataset = xr.open_dataset(filename)
+>>
+>>    #do your analyses on the data here
+>>
+>>    return dataset
+>>~~~
+>>
+> {: .solution}
+{: .challenge}
+
+> ## Reading data using Scipy's netCDF library
```

Comment:
I think this is the "old" NetCDF library that does not handle netcdf4. Perhaps we should use the one from unidata as the example instead? http://unidata.github.io/netcdf4-python/